internal/messagequeue/donthavetimeoutmgr.go

```diff
@@ -0,0 +1,270 @@
+	// The average latency of all pings
+	Latency() time.Duration
+}
+
+// dontHaveTimeoutMgr pings the peer to measure latency. It uses the latency to
+// set a reasonable timeout for simulating a DONT_HAVE message for peers that
+// don't support DONT_HAVE
+type dontHaveTimeoutMgr struct {
+	ctx                        context.Context
+	shutdown                   func()
+	peerConn                   PeerConnection
+	onDontHaveTimeout          func([]cid.Cid)
+	defaultTimeout             time.Duration
+	latencyMultiplier          int
+	maxExpectedWantProcessTime time.Duration
+
+	// All variables below here must be proctected by the lock
+	lk                    sync.RWMutex
+	started               bool
+	pending               map[cid.Cid]time.Time
+	timeout               time.Duration
+	checkForTimeoutsTimer *time.Timer
+}
+
+// newDontHaveTimeoutMgr creates a new dontHaveTimeoutMgr
+// onDontHaveTimeout is called when pending keys expire (not cancelled before timeout)
+func newDontHaveTimeoutMgr(ctx context.Context, pc PeerConnection, onDontHaveTimeout func([]cid.Cid)) *dontHaveTimeoutMgr {
+	return newDontHaveTimeoutMgrWithParams(ctx, pc, onDontHaveTimeout, dontHaveTimeout,
+		latencyMultiplier, maxExpectedWantProcessTime)
+}
+
+// newDontHaveTimeoutMgrWithParams is used by the tests
+func newDontHaveTimeoutMgrWithParams(ctx context.Context, pc PeerConnection, onDontHaveTimeout func([]cid.Cid),
+	defaultTimeout time.Duration, latencyMultiplier int,
+	maxExpectedWantProcessTime time.Duration) *dontHaveTimeoutMgr {
+
+	ctx, shutdown := context.WithCancel(ctx)
+	mqp := &dontHaveTimeoutMgr{
+		ctx:                        ctx,
+		shutdown:                   shutdown,
+		peerConn:                   pc,
+		pending:                    make(map[cid.Cid]time.Time),
+		timeout:                    defaultTimeout,
+		defaultTimeout:             defaultTimeout,
+		latencyMultiplier:          latencyMultiplier,
+		maxExpectedWantProcessTime: maxExpectedWantProcessTime,
+		onDontHaveTimeout:          onDontHaveTimeout,
+	}
+
+	return mqp
+}
+
+// Shutdown the dontHaveTimeoutMgr. Any subsequent call to Start() will be ignored
+func (dhtm *dontHaveTimeoutMgr) Shutdown() {
+	dhtm.shutdown()
+}
+
+// onShutdown is called when the dontHaveTimeoutMgr shuts down
+func (dhtm *dontHaveTimeoutMgr) onShutdown() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	// Clear any pending check for timeouts
+	if dhtm.checkForTimeoutsTimer != nil {
+		dhtm.checkForTimeoutsTimer.Stop()
+	}
+}
+
+// onStarted is called when the dontHaveTimeoutMgr starts.
+// It monitors for the context being cancelled.
+func (dhtm *dontHaveTimeoutMgr) onStarted() {
+	<-dhtm.ctx.Done()
+	dhtm.onShutdown()
+}
+
+// Start the dontHaveTimeoutMgr. This method is idempotent
+func (dhtm *dontHaveTimeoutMgr) Start() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	// Make sure the dont have timeout manager hasn't already been started
+	if dhtm.started {
+		return
+	}
+	dhtm.started = true
+
+	go dhtm.onStarted()
```

Comment:
We're not using these contexts for shutting down. We don't have to do this (as far as I know).

Reply:
In the constructor we create a context with cancel, and then keep a reference to the cancel.
This means that if the context is cancelled externally `cancel` will be called, and we can also explicitly call `cancel()` inside the class to shutdown.

Is there a better / more idiomatic way to do this?

Reply:
Well, the idiomatic way is to not take a context in the constructor. We're currently doing that for metrics.

Actually, in _this_ case, we can probably just drop the context from the constructor.

---

Note on idiomatic contexts: contexts should, in general, be used to manage _request_ lifetimes, not service lifetimes. Unfortunately, go contexts actually do two things:

1. They enable cancellation.
2. They carry "environment" variables.

We currently pass contexts down through constructors so we can pass through those environment variables (specifically, for logging). Ideally, we'd find a better way to do this.

Why not use contexts to cancel services? Canceling a context doesn't give you a way to _wait_ for something to stop. We've had a ton of problems where we cancel the main go-ipfs context, everything starts shutting down, then services start failing because other services they depend on have already shut down.

Also, contexts require this exact "spawn a goroutine to watch the context" pattern.

Reply:
Alternatively, we can just call this `closeAfterContext` or something like that. `onStarted` is where we call it, not what it does.

Reply:
Thanks for the detailed explanation ❤️ 

We need the context in order to do the ping, so for now I've just changed the method name to `closeAfterContext`. Let me know if there's a better alternative.